### ISSUE TITLE
Kinesis sink endpoint bug in AWS

### ISSUE THREAD

**Issue by MatissMa** [vectordotdev/vector#12021]
We are implementing Kinesis sink in AWS VPC.

Kinesis data stream is located in different AWS account and VPC than our EC2 instance with vector configuration. We have setup VPC peering between VPCs, we have adjusted security groups, added IAM roles with sts AssumeRole and configured route table accordingly. Our EC2 instance does not have internet access, so we are using VPC endpoints to access Kinesis data streams and sts. 
Kinesis endpoint is located on remote VPC, where Kinesis data streams are configured, but sts endpoint is local to our EC2 interneltess instance which have vector configured.

We have encountered an issue that if we configure Private DNS endpoint in `vector.toml` for Kinesis endpoint which is located in remote VPC, we received following error: 
![image](https://user-images.githubusercontent.com/43754605/160852958-bd092aa8-1895-4fc9-a963-2e112a6390e0.png)

We have these two VPC endpoints in use:
STS EP - IP 10.255.252.97
KinesIs EP - IP 10.75.2.73

EC2 instance with vector IP: 10.255.252.111

During the non-working configuration in tcpdump we do not see any traffic towards STS IP 10.255.252.97
![image](https://user-images.githubusercontent.com/43754605/160853994-c0c83e97-5265-495b-8855-3ace122b2074.png)

But we can see traffic towards Kinesis EP IP 10.75.2.73 via VPC peering
![image](https://user-images.githubusercontent.com/43754605/160854170-eff34a6f-234d-478c-a72a-5591e892955f.png)

The current workaround is to remove all endpoint configuration from `vector.toml` and this by default makes all STS commands to be sent to regional endpoint and all Kinesis commands to global endpoint **kinesis.eu-central-1.amazonaws.com**.

But since the EC2 does not have an internet access and cannot reach the global Kinesis endpoint, we have added DNS entry in `/etc/hosts` for global Kinesis endpoint to use the regional endpoint IP.
![image](https://user-images.githubusercontent.com/43754605/160855058-1d19c03e-108d-4877-aa5c-15fa16bbe20e.png)

When checking .pcap with this scenario we can verify that there is traffic to both STS endpoint and Kinesis endpoint:
![image](https://user-images.githubusercontent.com/43754605/160855215-b403f9f3-05f7-4bcd-94cd-77ece2f7aa49.png)

![image](https://user-images.githubusercontent.com/43754605/160855283-9b5fb3e2-3d39-46aa-9c1d-794a99542030.png)

Long story short, and our understanding is that if we configure endpoint for kinesis sink, everything seems to be sent to that endpoint (kinesis data stream and as well as sts), we can not specify vector to use local sts endpoint instead of confiugred Kinesis data stream endpoint. 
By removing endpoint configuration, adding region in the config, and adding in EC2 instance hosts file to resolve that global kinesis DNS to our peering Kinesis endpoint IP, then it works, because it resolves to private IP to which we have connectivity and sts goes via localy configured endpoint in that case.

**Comment by jszwedko** [comment]
Hi @MatissMa ,

Could you give this a try with the recently released `v0.21.0`? We swapped out the AWS SDK we are using for the official AWS one and I'm wondering if it handles this corectly.

**Comment by MatissMa** [comment]
Upgraded to version:
`vector 0.21.0 (aarch64-unknown-linux-gnu c1edb89 2022-04-14)`

Removed `/etc/hosts/` configuration, and for `vector.toml` added endpoint config and commented out region, as endpoint should be used when region is not (according to vector docummentation)
![image](https://user-images.githubusercontent.com/43754605/164266908-07b25250-b888-42ae-82c6-a91ab825a536.png)

This is the error that I receive:
![image](https://user-images.githubusercontent.com/43754605/164267173-da69ea78-3b72-4826-aa5e-76c7dc66430a.png)

And tried to add region as well the same time as endpoint is configured, received this error:
![image](https://user-images.githubusercontent.com/43754605/164267478-2570b190-581d-4f29-8bca-f111d62bdcdb.png)

Checked also tcpdump and it looks like it does not even send traffic towards configured kinesis endpoint nor local STS endpoint.

**Comment by jszwedko** [comment]
Hi @MatissMa !

With the new SDK, `region` is always required. Would you want to try setting that (along with `endpoint`)? We should fail at Vector boot time though rather than runtime though. I'll take a look into that.

**Comment by MatissMa** [comment]
This is the config (removed some sensitive info like account ID, and VPC endpoint DNS:
![image](https://user-images.githubusercontent.com/43754605/164281515-af2ab47b-eddd-4d64-b4f3-344c93a49418.png)


And this was the error I received with this configuration:
![image](https://user-images.githubusercontent.com/43754605/164280788-6e80bc91-b537-4654-896b-af1f0fa26be0.png)

Is there anything changed in syntax ? I can not even restart `vector.service` as I get error instantly. Thanks.

**Comment by jszwedko** [comment]
@MatissMa I think the `endpoint` needs `https://`. I'll try to verify that. Could you re-run with `RUST_BACKTRACE=1` and share the backtrace here?

**Comment by jszwedko** [comment]
@MatissMa thanks for helping troubleshoot. You can get the backtrace by exporting `RUST_BACKTRACE` as an environment variable with the value `1`. This should print a full backtrace when it panics.

**Comment by MatissMa** [comment]
hm, ok, so now I have a region and kinesis endpoint specified with https:// and I do not get "panicked", but this:
![image](https://user-images.githubusercontent.com/43754605/164288027-fce30c59-9ff4-4180-97b4-c65c01ec8e8e.png)

**Comment by jszwedko** [comment]
@MatissMa hmm, could you provide the must up-to-date configuration as text here? I can try to reproduce. There _should_ be a region being used there where it says `NoRegion`.

**Comment by MatissMa** [comment]
![image](https://user-images.githubusercontent.com/43754605/164289174-163a6f18-2e48-4ed4-ab51-e1fdfbb1362d.png)
This is the config for kinesis.sink at the moment.

**Comment by jszwedko** [comment]
Thanks @MatissMa ! I was able to reproduce. Looking now.

**Comment by slw07g** [comment]
Where can one get the nightly build from?

**Comment by slw07g** [comment]
Found it. :)

**Comment by slw07g** [comment]
@jszwedko I'm not the OP, but encountered the same issue and can confirm the nightly build solves it. Any idea when 0.21.1 will be avaiable?

**Comment by jszwedko** [comment]
> @jszwedko I'm not the OP, but encountered the same issue and can confirm the nightly build solves it. Any idea when 0.21.1 will be avaiable?

Thanks for confirming! We are targeting tomorrow for 0.21.1.

**Comment by MatissMa** [comment]
@jszwedko Tested today with updated version:
`vector 0.21.1`

![image](https://user-images.githubusercontent.com/43754605/165082918-52d73263-7317-444b-abe6-eba943810b06.png)

Exported RUST_BACKTRACE as well but nothing was printed out when kinesis panicked when vector.service was restarted. Any suggestions ? Thanks!

**Comment by MatissMa** [comment]
@jszwedko checked traces all good, thanks again!

**Comment by jszwedko** [comment]
Awesome, thanks for confirming @MatissMa !
